crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1335 @@
+use crate::entry::{Descriptor, Entry};
+use crate::sandboxed_tty_writer::SandboxedTTYWriter;
+use crate::wasi::wasi_snapshot_preview1::WasiSnapshotPreview1;
+use crate::wasi::{types, AsBytes, Errno, Result};
+use crate::WasiCtx;
+use crate::{clock, fd, path, poll};
+use log::{debug, error, trace};
+use std::cell::Ref;
+use std::convert::TryInto;
+use std::io::{self, Read, Seek, SeekFrom, Write};
+use std::ops::DerefMut;
+use wiggle_runtime::{GuestBorrows, GuestPtr};
+
+impl<'a> WasiSnapshotPreview1 for WasiCtx {
+    fn args_get<'b>(
+        &self,
+        mut argv: GuestPtr<'b, GuestPtr<'b, u8>>,
+        mut argv_buf: GuestPtr<'b, u8>,
+    ) -> Result<()> {
+        trace!("args_get(argv_ptr={:?}, argv_buf={:?})", argv, argv_buf);
+
+        let mut argv_buf_offset: types::Size = 0;
+        for arg in &self.args {
+            let arg_bytes = arg.as_bytes_with_nul();
+            let elems = arg_bytes.len().try_into()?;
+            argv_buf.as_array(elems).copy_from_slice(arg_bytes)?;
+            argv.write(argv_buf)?;
+            argv = argv.add(1)?;
+            argv_buf_offset = argv_buf_offset.checked_add(elems).ok_or(Errno::Overflow)?;
+            argv_buf = argv_buf.add(argv_buf_offset)?;
+        }
+
+        Ok(())
+    }
+
+    fn args_sizes_get(&self) -> Result<(types::Size, types::Size)> {
+        trace!("args_sizes_get");
+
+        let argc = self.args.len().try_into()?;
+        let mut argv_size: types::Size = 0;
+        for arg in &self.args {
+            let arg_len = arg.as_bytes_with_nul().len().try_into()?;
+            argv_size = argv_size.checked_add(arg_len).ok_or(Errno::Overflow)?;
+        }
+
+        trace!("     | *argc_ptr={:?}", argc);
+        trace!("     | *argv_buf_size_ptr={:?}", argv_size);
+
+        Ok((argc, argv_size))
+    }
+
+    fn environ_get<'b>(
+        &self,
+        mut environ: GuestPtr<'b, GuestPtr<'b, u8>>,
+        mut environ_buf: GuestPtr<'b, u8>,
+    ) -> Result<()> {
+        trace!(
+            "environ_get(environ={:?}, environ_buf={:?})",
+            environ,
+            environ_buf
+        );
+
+        let mut environ_buf_offset: types::Size = 0;
+        for e in &self.env {
+            let environ_bytes = e.as_bytes_with_nul();
+            let elems = environ_bytes.len().try_into()?;
+            environ_buf.as_array(elems).copy_from_slice(environ_bytes)?;
+            environ.write(environ_buf)?;
+            environ = environ.add(1)?;
+            environ_buf_offset = environ_buf_offset
+                .checked_add(elems)
+                .ok_or(Errno::Overflow)?;
```

Comment:
This looks like it may be buggy? I think `environ_buf_offset` may not be needed here because we're already offseting `environ_buf` on each iteration of the loop?

Reply:
@alexcrichton Right on, thanks for spotting that one! I believe I've fixed it in 5e75830, but if you could double check, that'd be awesome!